sys/clif/clif.c

```diff
@@ -0,0 +1,365 @@
+    [CLIF_PARAM_CT]        = LF_PARAM_CT_S,
+    [CLIF_PARAM_OBS]       = LF_PARAM_OBS_S
+};
+
+/* do not count extension param type */
+#define PARAMS_NUMOF (sizeof(_param_to_str) / sizeof(_param_to_str[0]))
+
+ssize_t clif_decode_link(clif_t *link, clif_param_t *params, unsigned params_len,
+                         char *buf, size_t maxlen)
+{
+    assert(buf && link);
+    char *pos = buf;
+    char *end = buf + maxlen;
+
+    char *link_start;
+    clif_param_t _dummy_param;
+    link->params = params;
+    link->target_len = 0;
+    link->params_len = 0;
+
+    ssize_t size = clif_get_link(pos, end - pos, &link_start);
+    if (size < 0) {
+        return CLIF_NOT_FOUND;
+    }
+    pos = link_start + size;
+
+    /* now get the target and params of the link */
+    size = clif_get_target(link_start, size, &link->target);
+    if (size < 0) {
+        return CLIF_NOT_FOUND;
+    }
+    link->target_len = size;
+    char *pos_in_link = link->target + size;
+
+    /* if there is no parameters array iterate all the buffer, if not until all
+     * the array is used */
+    while ((!params && pos_in_link < pos) || (params && link->params_len < params_len)) {
+        clif_param_t *param = params ? &params[link->params_len] : &_dummy_param;
+        ssize_t param_size = clif_get_param(pos_in_link, pos - pos_in_link,
+                                            param);
+        if (param_size < 0) {
+            break;
+        }
+        pos_in_link += param_size;
+        link->params_len++;
+    }
+
+    return pos - buf;
+}
+
+ssize_t clif_encode_link(const clif_t *link, char *buf, size_t maxlen)
+{
+    assert(link);
+    size_t pos = 0;
+    ssize_t res = 0;
+
+    res = clif_add_target(link->target, buf, maxlen);
+    if (res < 0) {
+        return res;
+    }
+    pos += res;
+
+    for (unsigned i = 0; i < link->params_len; i++) {
+        res = clif_add_param(&link->params[i], buf ? &buf[pos] : NULL,
+                             maxlen - pos);
+        if (res <= 0) {
+            return res;
+        }
+        pos += res;
+    }
+    return pos;
+}
+
+ssize_t clif_add_target(const char *target, char *buf, size_t maxlen)
+{
+    assert(target);
+
+    size_t pos = 0;
+    size_t target_len = strlen(target);
+
+    if (!buf) {
+        return target_len + 2; /* size after adding '<' and '>' */
+    }
+
+    if ((target_len + 2) > maxlen) {
+        return CLIF_NO_SPACE;
+    }
+
+    buf[pos++] = LF_PATH_BEGIN_C;
+
+    memcpy(&buf[pos], target, target_len);
+    pos += target_len;
+
+    buf[pos++] = LF_PATH_END_C;
+
+    return pos;
+}
+
+ssize_t clif_add_link_separator(char *buf, size_t maxlen)
+{
+    if (!buf) {
+        return 1;
+    }
+
+    if (maxlen < 1) {
+        return CLIF_NO_SPACE;
+    }
+
+    *buf = LF_LINK_SEPARATOR_C;
+    return 1;
+}
+
+ssize_t clif_add_param(clif_param_t *param, char *buf, size_t maxlen)
+{
+    assert(param && param->key);
+
+    param->key_len = strlen(param->key);
```

Comment:
Should we ignore the incoming value of key_len and assume that the key is 0-terminated? clif_get_param() sets key_len.

Reply:
Thanks, yes, makes sense to do it that way as well.